NEWS.md

```diff
@@ -293,6 +297,44 @@
 
 41. `tables()` is faster by default by excluding the size of character strings in R's global cache (which may be shared) and excluding the size of list column items (which also may be shared). `mb=` now accepts any function which accepts a `data.table` and returns a higher and better estimate of its size in bytes, albeit more slowly; e.g. `mb = utils::object.size`.
 
+42. Multiple improvements has been added to rolling functions. Request came from @gpierard who needed left aligned, adaptive, rolling max, [#5438](https://github.com/Rdatatable/data.table/issues/5438). There was no `frollmax` function yet. Adaptive rolling functions did not have support for `align="left"`. `frollapply` did not support `adaptive=TRUE`. Available alternatives were base R `mapply` or self-join using `max` and grouping `by=.EACHI`. As a follow up of his request, following features has been added:
```

Comment:
IMO this PR should be split up into those constituent parts:

 1. Add support for `frollapply(adaptive=TRUE)`
 1. Add support for `align="left"`
 1. Add `frollmax()`

It will be much easier to review those smaller PRs.

The parts about `partial=`, `give.names`, `hasNA`, and `Inf` in `froll{sum,mean}` could also be split off.

We can retain the one large NEWS item as here, but changes to code should be more piecemeal.

Reply:
I understand your point, and it would have been made like that if our PR merging process be smooth. Having a PR merged (let's say) once a month, you don't want to have 3 PRs waiting but rather single one.

Reply:
Understand that we have not had the ideal environment these past few years, I think starting from 1.15.99 we can restore discipline and get back into good habits.

Reply:
definitely, but considering 4 others PRs cascade from this one, it will be easier not to reorganize too much here

Reply:
Hmm, I don't follow.

Now we have `master -> A -> B -> C -> D`, I am proposing instead `master -> A0 -> A1 -> ... -> An -> A -> B -> C -> D`, i.e. this branch and downstreams unaffected, we just insert the split changes upstream of this one.

Any changes owing to review will need to be incorporated sequentially downstream, but that's no different than now.

Reply:
That could only work if A0...An diff would be exactly the same as A diff. And not sure if even then it would work smoothly, because history of B, C, D, E would contains different commits, only their sum would match.
And if diff wouldn't be exactly the same then possibly conflicts would have to be resolved in each of child branch.
Moreover splitting A into A0..An may be not trivial as there was a lot code refactoring in the meantime, not strictly related to particular features added.
Multiple pieces of this PR will be overwritten or reverted by child branches anyway, like next complete rewrite of frollapply.

Reply:
This PR is still _relatively_ small change comparing to all 5 PRs that I was proposing to review and merge at once :smile:

Reply:
Yes, it can be somewhat messy, but doable and worth it.

> code refactoring in the meantime, not strictly related to particular features added.

Again most suitable as its own PR.

To be clear, I am really quite loath to review the PR in its current state. If someone else wants to give Approval that's great.

Reply:
What are pretty good indicators in such cases
- if any existing tests have been changed
- if new code has code coverage
- what codeowners say :smile:

Reply:
https://graphite.dev/features/cli

the `gt split` tool would be a good fit for this situation 😉

tools listed here are designed to make the stacked PR process as painless as possible & remove some of the headache you mentioned re: sequential rebase of downstream PRs:

https://stacking.dev/

Reply:
given the current state here I'd be fine with that in this case -- commits that are cohesive in the actual code but don't necessarily pass CI, given we have a green commit here. appreciate the extra effort here!